draft-ietf-tls-esni.md

```diff
@@ -603,6 +606,26 @@ length.
 Backend servers (as described in {{server-behavior}}) MUST support the
 "ech_is_inner" extension.
 
+### GREASE PSK {#grease-psk}
+
+When offering ECH, the client is not permitted to advertise PSK identities in
+the ClientHelloOuter. However, it is useful to allow the client to add a fake
+"pre_shared_key" extension to the ClientHelloOuter whenever it advertises PSK
+identities in the ClientHelloInner. This provides cover traffic for session
+resumption with the backend server. When resuming a session with the client, the
+backend server sends a "pre_shared_key" extension in its ServerHello. This would
+appear to a network observer as if the the server were sending this extension
+without solicitation, which would violate the extension rules described in
+{{RFC8446}}. Sending a fake "pre_shared_key" extension in the ClientHelloOuter
+makes it appear to the network as if the extension were negotiated properly.
+
+The client generates the extension payload by constructing an `OfferedPsks`
+structure (see {{RFC8446}}, Section 4.2.11) as follows. For each PSK identity
+advertised in the ClientHelloInner, the client generates a random PSK identity
+with the same length. It also generates a random, 32-bit, unsinged integer to
+use as the `obfuscated_ticket_age`. Likewise, for each inner PSK binder, the
+client generates random string of the same length.
```

Comment:
Suppose the client does this, and then the server sends a ServerHello that claims to handshake with this PSK against ClientHelloOuter. What happens?

(This probably should be an error, i.e. the client should act as if it didn't offer PSKs. But it's ambiguous enough that we should write it down.)

Reply:
Good point -- I think noting this as an error and failing the handshake seems good enough.

Reply:
Added this line below:

> If the server replies with a "pre_shared_key" extension in its SeverHello, then
> the client MUST abort the handshake with a "handshake_failure" alert.

Reply:
Minor nitpick: Probably "unsupported_extension" or "illegal_parameter". "handshake_failure" is specifically if everyone followed the protocol correctly, just you were unable to pick a common set of parameter. Whereas the other alerts are variations of "you sent something invalid".

Reply:
I'll go with "illegal_parameter" :)